File: src/Library/assets/select-dropdown.js
/**
 * Alpine.data component for SelectDropdown
 * 
 * This file defines the interactive behavior for the SelectDropdown component.
 * It uses Alpine.js's data component system to provide reusable dropdown select functionality.
 * 
 * Architecture Overview:
 * - Registered as 'selectDropdown' Alpine.data component
 * - Manages dropdown state, search, and selection
 * - Supports single and multiple selection modes
 * - Provides keyboard navigation and search filtering
 * - Handles automatic positioning (drop up/down)
 * 
 * Dependencies:
 * - Alpine.js v3
 * 
 * Usage in templates:
 * x-data="selectDropdown(config)"
 */
document.addEventListener('alpine:init', () => {
    Alpine.data('selectDropdown', (config = {}) => ({
        // State management
        open: false,
        search: '',
        selected: config.value || (config.multiple ? [] : ''),
        highlighted: 0,
        
        // Configuration from component props
        options: config.options || [],
        multiple: config.multiple || false,
        searchable: config.searchable || false,
        clearable: config.clearable || false,
        placeholder: config.placeholder || 'Select...',
        maxHeight: parseInt(config.maxHeight) || 300,
        disabled: config.disabled || false,
        name: config.name || null,
        
        /**
         * Initialize component
         * Sets up initial state and watchers
         */
        init() {
            // Watch for selection changes to update form inputs
            this.$watch('selected', () => {
                this.updateFormInputs();
                // Dispatch custom event for external listeners
                this.$dispatch('select-change', { 
                    value: this.selected,
                    name: this.name 
                });
            });
            
            // Set initial form inputs
            this.updateFormInputs();
        },
        
        /**
         * Get filtered options based on search query
         * @returns {Array} Filtered options array
         */
        get filteredOptions() {
            if (!this.search) return this.options;
            
            const searchLower = this.search.toLowerCase();
            return this.options.filter(([value, label]) => 
                label.toLowerCase().includes(searchLower)
            );
        },
        
        /**
         * Get display label for selected value(s)
         * @returns {string} Label to display in the button
         */
        get selectedLabel() {
            if (this.multiple) {
                const labels = this.selected.map(v => {
                    const opt = this.options.find(([val]) => val === v);
                    return opt ? opt[1] : v;
                });
                return labels.length > 0 ? labels.join(', ') : this.placeholder;
            } else {
                if (!this.selected) return this.placeholder;
                const opt = this.options.find(([val]) => val === this.selected);
                return opt ? opt[1] : this.selected;
            }
        },
        
        /**
         * Check if there's a selection
         * @returns {boolean} True if something is selected
         */
        get hasSelection() {
            return this.multiple ? this.selected.length > 0 : !!this.selected;
        },
        
        
        /**
         * Select or deselect an option
         * @param {string} value - The option value to select
         */
        selectOption(value) {
            if (this.multiple) {
                const idx = this.selected.indexOf(value);
                if (idx > -1) {
                    this.selected.splice(idx, 1);
                } else {
                    this.selected.push(value);
                }
            } else {
                this.selected = value;
                this.open = false;
                this.search = '';
            }
        },
        
        /**
         * Clear all selections
         */
        clearSelection() {
            this.selected = this.multiple ? [] : '';
            this.search = '';
        },
        
        /**
         * Check if a value is selected
         * @param {string} value - The value to check
         * @returns {boolean} True if selected
         */
        isSelected(value) {
            return this.multiple 
                ? this.selected.includes(value) 
                : this.selected === value;
        },
        
        /**
         * Handle keyboard navigation
         * @param {KeyboardEvent} e - The keyboard event
         */
        handleKeydown(e) {
            // Open dropdown on Enter, Space, or ArrowDown when closed
            if (!this.open && ['Enter', ' ', 'ArrowDown'].includes(e.key)) {
                e.preventDefault();
                this.open = true;
                if (this.searchable) {
                    this.$nextTick(() => this.$refs.search?.focus());
                }
                return;
            }
            
            if (!this.open) return;
            
            switch(e.key) {
                case 'Escape':
                    e.preventDefault();
                    this.open = false;
                    this.$refs.button.focus();
                    break;
                    
                case 'ArrowDown':
                    e.preventDefault();
                    this.highlighted = Math.min(
                        this.highlighted + 1, 
                        this.filteredOptions.length - 1
                    );
                    this.scrollToHighlighted();
                    break;
                    
                case 'ArrowUp':
                    e.preventDefault();
                    this.highlighted = Math.max(this.highlighted - 1, 0);
                    this.scrollToHighlighted();
                    break;
                    
                case 'Enter':
                    e.preventDefault();
                    if (this.filteredOptions[this.highlighted]) {
                        this.selectOption(this.filteredOptions[this.highlighted][0]);
                    }
                    break;
                    
                case 'Tab':
                    // Let tab close the dropdown
                    this.open = false;
                    break;
            }
        },
        
        /**
         * Scroll to keep highlighted option in view
         */
        scrollToHighlighted() {
            this.$nextTick(() => {
                const container = this.$refs.optionsList;
                const highlighted = container?.querySelector(`[data-index="${this.highlighted}"]`);
                
                if (highlighted && container) {
                    const containerRect = container.getBoundingClientRect();
                    const highlightedRect = highlighted.getBoundingClientRect();
                    
                    if (highlightedRect.bottom > containerRect.bottom) {
                        container.scrollTop += highlightedRect.bottom - containerRect.bottom;
                    } else if (highlightedRect.top < containerRect.top) {
                        container.scrollTop -= containerRect.top - highlightedRect.top;
                    }
                }
            });
        },
        
        /**
         * Update hidden form inputs for form submission
         */
        updateFormInputs() {
            if (!this.name) return;
            
            // Find or create container for hidden inputs
            let container = this.$el.querySelector('[data-select-inputs]');
            if (!container) {
                container = document.createElement('div');
                container.setAttribute('data-select-inputs', '');
                container.style.display = 'none';
                this.$el.appendChild(container);
            }
            
            // Clear existing inputs
            container.innerHTML = '';
            
            // Create new inputs
            if (this.multiple) {
                this.selected.forEach(value => {
                    const input = document.createElement('input');
                    input.type = 'hidden';
                    input.name = this.name;
                    input.value = value;
                    container.appendChild(input);
                });
            } else if (this.selected) {
                const input = document.createElement('input');
                input.type = 'hidden';
                input.name = this.name;
                input.value = this.selected;
                container.appendChild(input);
            }
        },
        
        /**
         * Toggle dropdown open/closed
         */
        toggle() {
            if (this.disabled) return;
            
            this.open = !this.open;
            
            if (this.open && this.searchable) {
                this.$nextTick(() => this.$refs.search?.focus());
            }
        },
        
        /**
         * Handle clicks outside to close dropdown
         */
        handleClickOutside() {
            this.open = false;
            this.search = '';
        }
    }))
});

File: src/Library/forms/SelectDropdown.jl
"""
    @SelectDropdown

An enhanced dropdown select component with search functionality, keyboard navigation, and multiple selection support. Features type-ahead search, clear buttons, and full accessibility with Alpine.js integration.

# Props
- `options::Vector{Tuple{String,String}}`: Options as (value, label) tuples
- `placeholder::Union{String,Nothing}`: Placeholder text (default: "Select...")
- `searchable::Bool`: Enable search functionality (default: `false`)
- `multiple::Bool`: Enable multiple selection (default: `false`)
- `clearable::Bool`: Enable clear button to reset selection (default: `false`)
- `max_height::String`: Maximum height of dropdown (default: "300px")
- `size::Union{Symbol,String}`: Component size (`:xs`, `:sm`, `:base`, `:lg`, `:xl`) (default: `:base`)
- `state::Union{Symbol,String}`: Component state (`:default`, `:error`, `:success`) (default: `:default`)
- `name::Union{String,Nothing}`: Form field name (optional)
- `value::Union{String,Vector{String},Nothing}`: Selected value(s) (optional)
- `disabled::Bool`: Whether component is disabled (default: `false`)
- `required::Bool`: Whether field is required (default: `false`)
- `id::Union{String,Nothing}`: Component ID (optional)

# Requirements
This component requires Alpine.js and Alpine Anchor for intelligent positioning:

```html
<script defer src="https://cdn.jsdelivr.net/npm/@alpinejs/anchor@latest/dist/cdn.min.js"></script>
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3/dist/cdn.min.js"></script>
```

**Browser Compatibility:** Modern browsers with ES6 support  
**Dependencies:** Tailwind CSS for styling classes

**Note:** JavaScript assets are automatically loaded via `@__once__` for optimal performance.

# Accessibility
**ARIA Patterns:** Uses `role="listbox"` with proper option roles, `aria-expanded` state, and `aria-selected` for selections.

**Keyboard Navigation:** Space/Enter to open, Arrow keys to navigate, Escape to close, type-ahead search support.

**Screen Reader Support:** Selection changes, option count, and search functionality are announced. Works with form validation.

**Visual Design:** High contrast focus indicators and touch-friendly spacing on mobile devices.
"""
@component function SelectDropdown(;
    options::Vector{Tuple{String,String}} = Tuple{String,String}[],
    placeholder::Union{String,Nothing} = "Select...",
    searchable::Bool = false,
    multiple::Bool = false,
    clearable::Bool = false,
    max_height::String = "300px",
    size::Union{Symbol,String} = :base,
    state::Union{Symbol,String} = :default,
    name::Union{String,Nothing} = nothing,
    value::Union{String,Vector{String},Nothing} = nothing,
    disabled::Bool = false,
    required::Bool = false,
    id::Union{String,Nothing} = nothing,
    _hash = hash(time_ns()),
    attrs...,
)
    # Load JavaScript for dropdown functionality
    @__once__ begin
        @script @text SafeString(
            read(joinpath(@__DIR__, "../assets/select-dropdown.js"), String),
        )
    end
    # Convert to symbols
    size_sym = Symbol(size)
    state_sym = Symbol(state)

    # Generate unique ID if not provided
    component_id = isnothing(id) ? "select-dropdown-$(_hash)" : id
    dropdown_id = "$(component_id)-dropdown"

    # Helper to escape JavaScript strings  
    escape_js(s::String) = replace(replace(s, "\\" => "\\\\"), "'" => "\\'")

    # Create configuration object for Alpine component
    # Using JSON-like syntax for better reliability
    options_json = "["
    for (i, (val, label)) in enumerate(options)
        if i > 1
            options_json *= ", "
        end
        options_json *= "['$(escape_js(val))', '$(escape_js(label))']"
    end
    options_json *= "]"

    # Build configuration string
    config_parts = String[]
    push!(config_parts, "options: $options_json")
    push!(config_parts, "multiple: $(multiple ? "true" : "false")")
    push!(config_parts, "searchable: $(searchable ? "true" : "false")")
    push!(config_parts, "clearable: $(clearable ? "true" : "false")")
    push!(config_parts, "disabled: $(disabled ? "true" : "false")")
    push!(config_parts, "maxHeight: '$(escape_js(max_height))'")

    if !isnothing(placeholder)
        push!(config_parts, "placeholder: '$(escape_js(placeholder))'")
    end

    if !isnothing(name)
        push!(config_parts, "name: '$(escape_js(name))'")
    end

    # Handle initial value
    if !isnothing(value)
        if multiple
            value_json = "[" * join(["'$(escape_js(v))'" for v in value], ", ") * "]"
            push!(config_parts, "value: $value_json")
        else
            push!(config_parts, "value: '$(escape_js(value))'")
        end
    end

    alpine_config = SafeString("{" * join(config_parts, ", ") * "}")

    # Size classes (matching Input component)
    size_classes = (
        xs = "px-2.5 py-1.5 text-xs",
        sm = "px-3 py-2 text-sm",
        base = "px-4 py-2.5 text-base",
        lg = "px-5 py-3 text-lg",
        xl = "px-6 py-3.5 text-xl",
    )

    state_classes = (
        default = "border-slate-300 focus:border-indigo-500 focus:ring-indigo-500/20 dark:border-slate-600 dark:focus:border-indigo-400",
        error = "border-rose-300 focus:border-rose-500 focus:ring-rose-500/20 dark:border-rose-600 dark:focus:border-rose-400",
        success = "border-emerald-300 focus:border-emerald-500 focus:ring-emerald-500/20 dark:border-emerald-600 dark:focus:border-emerald-400",
    )

    size_class = get(size_classes, size_sym, size_classes.base)
    state_class = get(state_classes, state_sym, state_classes.default)
    disabled_class = disabled ? "opacity-60 cursor-not-allowed" : "cursor-pointer"

    # Build the component
    @div {
        var"x-data" = SafeString("selectDropdown($alpine_config)"),
        var"@keydown" = "handleKeydown(\$event)",
        class = "relative",
        attrs...,
    } begin
        # Hidden inputs container will be managed by Alpine component
        @div {var"data-select-inputs" = "", style = "display: none"}

        # Wrapper for button and clear button
        @div {class = "relative"} begin
            # Dropdown trigger button
            @button {
                type = "button",
                var"x-ref" = "button",
                var"@click" = "toggle()",
                var":aria-expanded" = "open.toString()",
                var"aria-haspopup" = "listbox",
                var"aria-controls" = dropdown_id,
                disabled = disabled,
                required = required,
                class = "w-full flex items-center justify-between rounded-md border bg-white dark:bg-slate-900 text-slate-900 dark:text-slate-100 focus:outline-none focus:ring-2 focus:ring-offset-0 transition-colors duration-150 hover:border-slate-400 dark:hover:border-slate-500 $size_class $state_class $disabled_class",
            } begin
                @span {
                    var"x-text" = "selectedLabel",
                    var":class" = "{ 'text-slate-500 dark:text-slate-400': !hasSelection }",
                    class = clearable ? "pr-12" : "pr-8",
                } begin
                    if !isnothing(placeholder)
                        @text placeholder
                    end
                end

                # Dropdown arrow
                @svg {
                    class = "absolute right-3 h-5 w-5 text-slate-400 transition-transform duration-150 pointer-events-none",
                    var":class" = "{ 'rotate-180': open }",
                    xmlns = "http://www.w3.org/2000/svg",
                    viewBox = "0 0 20 20",
                    fill = "currentColor",
                } begin
                    @path {
                        var"fill-rule" = "evenodd",
                        d = "M5.293 7.293a1 1 0 011.414 0L10 10.586l3.293-3.293a1 1 0 111.414 1.414l-4 4a1 1 0 01-1.414 0l-4-4a1 1 0 010-1.414z",
                        var"clip-rule" = "evenodd",
                    }
                end
            end

            # Clear button (outside main button)
            if clearable
                @button {
                    type = "button",
                    var"@click.stop" = "clearSelection()",
                    var"x-show" = "hasSelection",
                    var"x-transition:enter" = "transition ease-out duration-150",
                    var"x-transition:enter-start" = "opacity-0 scale-75",
                    var"x-transition:enter-end" = "opacity-100 scale-100",
                    var"x-transition:leave" = "transition ease-in duration-100",
                    var"x-transition:leave-start" = "opacity-100 scale-100",
                    var"x-transition:leave-end" = "opacity-0 scale-75",
                    class = "absolute right-10 top-1/2 -translate-y-1/2 p-1 rounded-md hover:bg-slate-100 dark:hover:bg-slate-800 focus:outline-none focus:ring-2 focus:ring-indigo-500/30",
                    var"aria-label" = "Clear selection",
                } begin
                    @svg {
                        class = "h-4 w-4 text-slate-500 dark:text-slate-400",
                        xmlns = "http://www.w3.org/2000/svg",
                        viewBox = "0 0 20 20",
                        fill = "currentColor",
                    } begin
                        @path {
                            var"fill-rule" = "evenodd",
                            d = "M4.293 4.293a1 1 0 011.414 0L10 8.586l4.293-4.293a1 1 0 111.414 1.414L11.414 10l4.293 4.293a1 1 0 01-1.414 1.414L10 11.414l-4.293 4.293a1 1 0 01-1.414-1.414L8.586 10 4.293 5.707a1 1 0 010-1.414z",
                            var"clip-rule" = "evenodd",
                        }
                    end
                end
            end
        end

        # Dropdown panel
        @div {
            var"x-show" = "open",
            var"x-anchor.bottom-start.offset.4" = SafeString("\$refs.button"),
            var"x-transition:enter" = "transition ease-out duration-150",
            var"x-transition:enter-start" = "opacity-0 transform scale-95",
            var"x-transition:enter-end" = "opacity-100 transform scale-100",
            var"x-transition:leave" = "transition ease-in duration-100",
            var"x-transition:leave-start" = "opacity-100 transform scale-100",
            var"x-transition:leave-end" = "opacity-0 transform scale-95",
            var"@click.away" = "handleClickOutside()",
            id = dropdown_id,
            class = "absolute z-50 w-full rounded-lg bg-white dark:bg-slate-900 shadow-[0_4px_12px_rgba(0,0,0,0.08)] ring-1 ring-slate-200 dark:ring-slate-700 overflow-hidden",
            role = "listbox",
            var"aria-label" = placeholder,
        } begin
            # Search input (if searchable)
            if searchable
                @div {class = "p-2 border-b border-slate-200 dark:border-slate-700"} begin
                    @input {
                        type = "text",
                        var"x-ref" = "search",
                        var"x-model" = "search",
                        var"@click.stop" = "",
                        placeholder = "Search...",
                        class = "w-full px-3 py-2 text-sm rounded-md border border-slate-300 dark:border-slate-600 bg-white dark:bg-slate-900 text-slate-900 dark:text-slate-100 placeholder:text-slate-400 dark:placeholder:text-slate-500 focus:outline-none focus:ring-2 focus:ring-indigo-500/20 focus:border-indigo-500",
                    }
                end
            end

            # Options list
            @div {
                class = "overflow-y-auto",
                style = "max-height: $max_height",
                var"x-ref" = "optionsList",
            } begin
                @template {
                    var"x-for" = "(option, index) in filteredOptions",
                    var":key" = "option[0]",
                } begin
                    @button {
                        type = "button",
                        var"@click" = "selectOption(option[0])",
                        var":class" = """{
                            'bg-indigo-50 dark:bg-indigo-900/20': highlighted === index,
                            'bg-indigo-100 dark:bg-indigo-900/40': isSelected(option[0])
                        }""",
                        var"@mouseenter" = "highlighted = index",
                        var":data-index" = "index",
                        class = "w-full px-4 py-2.5 text-left text-sm hover:bg-slate-50 dark:hover:bg-slate-800 focus:outline-none focus:bg-slate-50 dark:focus:bg-slate-800 transition-colors duration-150",
                        role = "option",
                        var":aria-selected" = "isSelected(option[0])",
                    } begin
                        @div {class = "flex items-center"} begin
                            if multiple
                                @div {class = "mr-3"} begin
                                    @div {
                                        class = "h-4 w-4 rounded border-2 transition-colors duration-150",
                                        var":class" = """{
                                            'border-indigo-500 bg-indigo-500': isSelected(option[0]),
                                            'border-slate-300 dark:border-slate-600': !isSelected(option[0])
                                        }""",
                                    } begin
                                        @svg {
                                            var"x-show" = "isSelected(option[0])",
                                            class = "h-3 w-3 text-white",
                                            fill = "none",
                                            viewBox = "0 0 24 24",
                                            stroke = "currentColor",
                                        } begin
                                            @path {
                                                var"stroke-linecap" = "round",
                                                var"stroke-linejoin" = "round",
                                                var"stroke-width" = "3",
                                                d = "M5 13l4 4L19 7",
                                            }
                                        end
                                    end
                                end
                            end
                            @span {
                                var"x-text" = "option[1]",
                                class = "text-slate-900 dark:text-slate-100",
                            }
                        end
                    end
                end

                # No results message
                @div {
                    var"x-show" = "search && filteredOptions.length === 0",
                    class = "px-4 py-8 text-center text-sm text-slate-500 dark:text-slate-400",
                } "No options found"
            end
        end
    end
end

@deftag macro SelectDropdown end
